examples/mciso_advanced.py

```diff
@@ -344,16 +347,91 @@ def march(self) -> int:
                 if self.m[i + 1, j + 1, k + 1] > 1: id |= 64
                 if self.m[i, j + 1, k + 1] > 1: id |= 128
 
-            E = [ti.Vector(_) + .5 for _ in [self.E2, self.E3][self.dim - 2]]
-
-            for k in range(self.et.shape[1]):
-                if self.et[id, k][0] == -1:
+            for m in range(self.et.shape[1]):
+                if self.et[id, m][0] == -1:
                     break
 
                 n = ti.atomic_add(r_n, 1)
                 for l in ti.static(range(self.dim)):
-                    e = self.et[id, k][l]
-                    self.r[n, l] = I + list_subscript(E, e)
+                    e = self.et[id, m][l]
+                    R = float(I)
+
+                    if ti.static(self.dim == 2):
+                        i, j = I
+                        # (.5, 0), (0, .5), (1, .5), (.5, 1)
+                        if e == 0:
+                            p = (1 - self.m[i, j]) / (self.m[i + 1, j] -
+                                                      self.m[i, j])
+                            R = [i + p, j]
+                        elif e == 1:
+                            p = (1 - self.m[i, j]) / (self.m[i, j + 1] -
+                                                      self.m[i, j])
+                            R = [i, j + p]
+                        elif e == 2:
+                            p = (1 - self.m[i + 1, j]) / (
+                                self.m[i + 1, j + 1] - self.m[i + 1, j])
+                            R = [i + 1, j + p]
+                        elif e == 3:
+                            p = (1 - self.m[i, j + 1]) / (
+                                self.m[i + 1, j + 1] - self.m[i, j + 1])
+                            R = [i + p, j + 1]
+                    else:
+                        i, j, k = I
+                        # (.5, 0, 0), (1, .5, 0), (.5, 1, 0), (0, .5, 0)
```

Comment:
Just an idea, it might be easier for you to maintain if these coordinate offsets are put into a field. You can then index them with `e`?